### PR TITLE
Cleanup unmaintainable parallel logic with workers

### PR DESCRIPTION
**Pull request by tjwald**
Relates to [Issue](https://github.com/dotnet/runtime/issues/53593)
This PR is a continuation of the previous [PR](https://github.com/dotnet/runtime/pull/76949)

This PR uses worker objects and worker bodies to unify the implementation of 3 out of the 4 core logics in Parallel:
1. ```For32```
2. ```For64```
3. ```Foreach``` (Both Orderable and Partitioner)

The only one that is not covered is ```Invoke``` that had a completely separate implementation.

### CODE REVIEW

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs [hunk 1]
```diff
@@ -9,13 +9,19 @@
...
+using System.Numerics;
+using System.Runtime.ExceptionServices;
 
 namespace System.Threading.Tasks
 {
+    internal sealed class Box<T> where T: class
+    {
+        public T? Value { get; set; }
+    }
+
```
Comment (teo-tsirpanis) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs:24]: How about you use `System.Runtime.CompilerServices.StrongBox<T>` instead?

**Comment by joperezr** [comment]
cc: @stephentoub

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs (hunk 1 above)
Comment (tjwald) [reply src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs:24]: Cool! didn't know it existed - I assumed there had to be one already implemented.

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs [hunk 2]
```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
...
+    #endregion
+
+    #region Worker Bodies Factories
+
+    internal interface IWorkerBodyFactory<in TValue>
+    {
+        IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags);
+    }
+
+    internal sealed class WorkerWithSimpleBodyFactory<TValue>: IWorkerBodyFactory<TValue>
```
Comment (tjwald) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs:277]: I can change all the factories to structs if that is desirable - there is currently no inheritance. 
This means I could change the functions receiving them to be generic over the factory and (theoretically) reduce allocation.

Should I do this?

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs [hunk 3]
```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
...
+            _body = body;
+        }
+
+        public void Body(TValue index)
+        {
+            _body(index);
+        }
+    }
+
+    internal abstract class WorkerBodyWithIndex<TValue, TIndex> : IWorkerBodyWithIndex<TValue, TIndex>
```
Comment (tjwald) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs:154]: with little code duplication in the worker classes I can change them classes to structs and reduce allocations if that is desirable. 
Thoughts?

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs [hunk 4]
```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
...
+
+        public IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags)
+            => new WorkerWithEverything<TValue, TIndex, TLocal>(sharedPStateFlags, _bodyWithLocal, _localInit, _localFinally);
+    }
+
+    #endregion
+
+    #region Workers
+
+    internal interface IWorkerFactory<in TSource, TInput, TValue, out TIndex> where TIndex: INumber<TIndex>
```
Comment (tjwald) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs:355]: This interface started as a factory only interface but now contains more loop run related metadata.
Is there a better name for this?

**Comment by tjwald** [comment]
Is there an automatic formatter that complies with the runtime repo guidelines?

**Comment by tjwald** [comment]
Where and what docs should I add in the code?

**Comment by dakersnar** [comment]
@tjwald is this PR a superset (or direct improvement) of https://github.com/dotnet/runtime/pull/76949? If so, we should be able to close https://github.com/dotnet/runtime/pull/76949, correct?

**Comment by tjwald** [comment]
> @tjwald is this PR a superset (or direct improvement) of #76949? If so, we should be able to close #76949, correct?

This is a superset. 
I didn't know if you would like this solution, so I left the old PR if we would want to just merge that.
But I like this PR's outcome and if you agree we should close the old PR.

**Comment by dakersnar** [comment]
Sounds good, thanks for the details. Expect a deeper review of both PRs in early January, and feel free to ping me if this falls through the cracks.

**Comment by tjwald** [comment]
@dakersnar Hi! any news? I didn't remined earlier since I was busy.

**Comment by dakersnar** [comment]
Thanks for the reminder, and sorry for the delay. This is on my agenda; I'll take a look early next week.

**Comment by stephentoub** [comment]
@tjwald, thank you for trying to tackle this, and apologies for taking so long to getting around to looking at it.  I very much appreciate your work here and your interest in contributing.

That said, I'm concerned about a couple of things.  First and foremost (and having absolutely nothing to do with your work here), there is a lot of code involved here that's been working fine for a decade.  The issue that was opened about consolidating logic wasn't a request from the .NET team but rather a comment from someone in the community that it could be cleaned up.  And while it's certainly nice to try to consolidate, to ensure there aren't any regressions we'd need to go through the changes line-by-line with a fine-toothed comb to confirm that semantics are being maintained 100%.  With the long-term maintenance responsibility on our shoulders, this is one of those times where it's a lot harder for us to accept an external contribution, because the time it takes to review/compare line-by-line and validate that nothing has gone wrong accidentally is a lot more than the time it takes to just do the work.  Second, I see in the changes that new instance interface methods and new instance virtual methods are being introduced; that's an example of the kind of change that can have a negative performance implication rather than functional regressions like I was concerned about in my first point.  Third, from just a quick skim I do see places where there are functional problems, e.g. making the ETW provider generic and leading to different event signatures based on which caller is being used.

This issue shouldn't have been marked as "help wanted".  That's our bad, and I apologize for that and that it's ended up wasting your time.  I'm going to close this PR and try my hand at doing the consolidation myself.  I'll then compare what I came up with with what you have here... if they end up being close enough, I'll cherry-pick some of your commits.  Either way, I can tag you when I open a PR for it in case you'd like to review.

Thank you, again, and apologies, again.  If you'd like some help choosing a better issue on which to contribute, I'd be happy to help.

**Comment by tjwald** [comment]
@stephentoub I perfectly understand, and I am happy that the minimal change was done and merged! 
If there is anything that is around this level of difficulty and time constraints then I would love to try and contribute!
However I couldn't find anything on my own so if you could help me I would really appreciate it.

**Pull request closed without merging**